fw/sock_clnt.c

```diff
@@ -428,56 +197,57 @@ tfw_sk_prepare_xmit(struct sock *sk, struct sk_buff *skb, unsigned int mss_now,
 		goto err_purge_tcp_write_queue;
 	}
 
-	*nskbs = UINT_MAX;
 	h2_mode = TFW_CONN_PROTO(conn) == TFW_FSM_H2;
-	if (h2_mode)
-		return tfw_h2_sk_prepare_xmit(sk, skb, mss_now, limit, nskbs);
+	if (!h2_mode)
+		return 0;
 
-	return r;
+	h2 = tfw_h2_context(conn);
 
-err_purge_tcp_write_queue:
 	/*
-	 * Leave encrypted segments in the retransmission rb-tree,
-	 * but purge the send queue on unencrypted segments.
+	 * We use two different strategies here:
+	 * - The main strategy is to calculate count of bytes that we can
+	 *   send to the client immediately, according to mss and available
+	 *   TCP CWND. We try to prepare HTTP frame which size is greater
+	 *   or less than mss to prevent HOL problem. The fact that we don't
+	 *   send everything at once allows the higher priority data to be
+	 *   sent before the lower priority data, as we select a new highest
+	 *   priority data to send with each acknowledgement.
+	 * - The second strategy is used when mss is less or equal to frame
+	 *   header size (9 byte, it is possible with very low MTU) or if we
+	 *   finish TCP connection with FIN and should send all available data.
+	 *   In this case we don't care about efficiency and just make frames
+	 *   ignoring mss and CWND limits.
 	 */
-	tcp_write_queue_purge(sk);
-	return r;
-}
-
-static int
-tfw_sk_write_xmit(struct sock *sk, struct sk_buff *skb, unsigned int mss_now,
-		  unsigned int limit, unsigned int nskbs)
-{
-	TfwConn *conn = sk->sk_user_data;
-	TfwH2Ctx *h2;
-	TfwHPackETbl *tbl;
-	unsigned short flags;
-	bool h2_mode;
-	int r = 0;
-
-	assert_spin_locked(&sk->sk_lock.slock);
-	/* Should be checked early in `tfw_sk_prepare_xmit`. */
-	BUG_ON(!conn);
+	mss_now = (mss_now > TLS_MAX_PAYLOAD_SIZE + TLS_MAX_OVERHEAD ?
+		TLS_MAX_PAYLOAD_SIZE : mss_now - TLS_MAX_OVERHEAD);
+	if (mss_now <= FRAME_HEADER_SIZE) {
+		mss_now = UINT_MAX;
+		with_limit = false;
+	}
+	if (with_limit) {
+		cwnd_avail = (tp->snd_cwnd > tp->packets_out ?
+			tp->snd_cwnd - tp->packets_out : 0);
+		cwnd_avail_in_bytes = cwnd_avail * mss_now;
```

Comment:
Rework this place, but you also not right here. `tcp_cwnd_test` returns `min(halfcwnd, cwnd - in_flight)` on each iteration of the loop for calculation count of bytes that can be sent per ONE loop iteration. Since we call this function on each loop iteration it returns 0 only when cwnd - in_flight became equal to zero! So we can use cwnd - in_flight for our purpose, but also we should take into account len of socket write queue, because we can push a lot of skbs on previous iteration if headers are big (we ignore cwnd when we make headers) or if there are some special frames were sent. We also cant use tcp_mss_split_point because it deal with skb. We can use it only if we make frame for each skb (use skb->len as a limit of frame length), but it i think, that it is bad, since we should use tls encryption for each skb in this case. So i think that rough calculation is better.